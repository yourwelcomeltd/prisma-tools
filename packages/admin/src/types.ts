import React from 'react';
import { Column, UseSortByColumnOptions } from 'react-table';
import { AdminSchema, AdminSchemaField, AdminSchemaModel } from '@paljs/types';
import Language from './PrismaTable/language';
import { DynamicTableProps } from './PrismaTable/dynamicTable';
import { RegisterOptions } from 'react-hook-form/dist/types/validator';

export type { AdminSchema, AdminSchemaField, AdminSchemaModel };

export type Columns = Record<
  'boolean' | 'number' | 'enum' | 'DateTime' | 'object' | 'string' | 'list' | 'json',
  Column<Record<string, any>> & UseSortByColumnOptions<any>
>;

export interface InputProps {
  field: AdminSchemaField;
  value: any;
  data: any;
  disabled: boolean;
}

export type GetColumns = (field: AdminSchemaField, model?: AdminSchemaModel) => Columns;
export type GetColumnsPartial = (field: AdminSchemaField, model?: AdminSchemaModel) => Partial<Columns>;
export type FormInputs = Record<
  'Default' | 'Editor' | 'Enum' | 'Object' | 'Date' | 'Boolean' | 'Upload',
  React.FC<InputProps>
>;

export interface ContextProps extends RequireContextProps, SameProps {
  schema: AdminSchema;
  children?: React.ReactNode;
}

interface RequireContextProps {
  pagesPath: string;
  pageSize: number;
  pageSizeOptions: number[];
  paginationOptions: number;
  lang: typeof Language;
  dir: 'rtl' | 'ltr';
}

interface SameProps {
  actions?: ('create' | 'update' | 'delete')[];
  useSet?: boolean;
  tableColumns?: GetColumnsPartial;
  formInputs?: Partial<FormInputs>;
  inputValidation?: Record<string, Record<string, RegisterOptions>>;
  push: (url: string) => void;
  query: Record<string, any>;
  onSelect?: (values: any[]) => void;
  onCancelCreate?: (options: { model: string; setCreateModal: (state: boolean) => void }) => void;
  onSaveCreate?: (options: {
    model: string;
    setCreateModal: (state: boolean) => void;
    refetchTable: (options?: any) => void;
  }) => void;
  onCancelUpdate?: (options: { model: string }) => void;
  onSaveUpdate?: (options: { model: string; refetchTable: (options?: any) => void }) => void;
  valueHandler?: (value: string, field?: AdminSchemaField, create?: boolean) => any;
  actionButtons?: {
    Add?: React.FC;
    Update?: React.FC<{ id: any }>;
    Delete?: React.FC<{ id: any }>;
  };
  defaultOrderBy?: Record<string, Record<string, 'asc' | 'desc' | { sort: 'asc' | 'desc'; nulls: 'last' | 'first' }>[]>;
}

export interface ModelTableProps extends Partial<Omit<RequireContextProps, 'lang'>>, SameProps {
  model: string;
  language?: Partial<typeof Language>;
  children?: DynamicTableProps['children'];
}

export interface TableParentRecord {
  name: string;
  value: any;
  field: string;
  updateRecord?: () => Promise<any>;
}
