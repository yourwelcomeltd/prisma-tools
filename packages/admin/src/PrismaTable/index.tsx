import React from 'react';
import { useQuery } from '@apollo/client';

import Spinner from '../components/Spinner';
import DynamicTable from './dynamicTable';
import { GET_SCHEMA } from '../SchemaQueries';
import { ModelTableProps, ContextProps } from '../index';
import { TableContext, defaultSettings } from './Context';
import defaultLanguage from './language';

const PrismaTable: React.FC<ModelTableProps> = ({ children, language, model, ...rest }) => {
  const { data, loading } = useQuery<{ getSchema: ContextProps['schema'] }>(GET_SCHEMA);
  if (loading) return <Spinner />;
  const mergedLanguage = { ...defaultLanguage, ...language };
  return (
    <TableContext.Provider
      value={{
        schema: data?.getSchema ?? {
          models: [],
          enums: [],
        },
        ...({ ...defaultSettings, ...rest } as any),
        lang: mergedLanguage,
      }}
    >
      <DynamicTable model={model}>{children}</DynamicTable>
    </TableContext.Provider>
  );
};

export { PrismaTable, TableContext };
export * from './Table/Filters';
export * from './Form/Inputs';
