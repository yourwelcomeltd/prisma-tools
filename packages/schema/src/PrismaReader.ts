import { existsSync, readFileSync } from 'fs';
import { Field } from '@paljs/types';
import { log, chalk } from '@paljs/display';

export class PrismaReader {
  data: string;

  constructor(protected path: string) {
    this.checkIfSchemaExit();
    this.data = readFileSync(path, { encoding: 'utf-8' });
  }

  get models() {
    return this.data.match(/\n(model(\s)[\s\S]*?})\r?\n/g);
  }

  get enums() {
    return this.data.match(/\n(enum(\s)[\s\S]*?})\r?\n/g);
  }

  getModelDocumentation(modelName: string) {
    return this.data
      .split(/\n/)
      .reverse()
      .filter((_item, index, array) => array.indexOf(`model ${modelName} {`) < index)
      .filter((_item, index, array) => array.indexOf(`}`) > index)
      .filter(Boolean)
      .join('\n');
  }

  blockLines(block: string) {
    return block.split(/\n/).filter((v) => v);
  }

  checkIfSchemaExit() {
    if (!existsSync(this.path)) {
      log.error(`Error: ${chalk.blue('schema.prisma')} file not found in ${chalk.blue(this.path)}`);
      process.exit();
    }
  }

  getType(line: string[]) {
    return line[1].replace('?', '').replace('[]', '');
  }

  getKind(type: string) {
    return this.data.includes(`enum ${type} `) ? 'enum' : this.data.includes(`model ${type} `) ? 'object' : 'scalar';
  }

  getClassName(lines: string[]) {
    return this.lineArray(lines[0])[1];
  }

  lineArray(line: string) {
    return line
      .replace(/[\n\r]/g, '')
      .split(' ')
      .filter(Boolean);
  }

  getMap(line: string) {
    const value = line.match(/@map\((.*?)\)/);
    if (value) {
      return value[1].replace(/name/, '').replace(':', '').replace(' ', '').replace(/"/g, '');
    }
    return undefined;
  }

  getRelation(line: string) {
    const relationString = line.match(/@relation\((.*?)\)/);
    if (relationString) {
      const relation: Field['relation'] = {};
      const name = relationString[1].match(/"(\w+)"/);
      if (name) {
        relation.name = name[1];
      }
      ['fields', 'references'].forEach((item) => {
        const pattern = new RegExp(`${item}:[\\s\\S]\\[(.*?)\\]`);
        const values = relationString[1].match(pattern);
        if (values) {
          const asArray = values[1]
            .replace(/ /g, '')
            .split(',')
            .filter((v) => v);
          if (asArray.length > 0) {
            relation[item as 'fields' | 'references'] = asArray;
          }
        }
      });
      return relation;
    }
    return undefined;
  }
}
