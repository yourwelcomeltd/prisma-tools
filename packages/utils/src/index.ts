export * from './dmmf';
export * from './schemaInputs';
export { getDMMF, formatSchema } from '@prisma/internals';
