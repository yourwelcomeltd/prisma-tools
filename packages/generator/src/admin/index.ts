import { writeFileSync } from 'fs';
import { format } from 'prettier';
import { GeneratorOptions, AdminPagesOptions, SchemaObject } from '@paljs/types';
import { createGraphql } from './createGraphql';
import { mergeSchema } from './mergeSchema';
import { ConvertSchemaToObject } from '@paljs/schema';
import { createFile } from './createFile';

const defaultOptions: GeneratorOptions = {
  prismaName: 'prisma',
  output: './src/graphql',
  excludeFields: [],
  excludeModels: [],
  excludeFieldsByModel: {},
  excludeQueriesAndMutations: [],
  excludeQueriesAndMutationsByModel: {},
};

export class UIGenerator {
  schema: SchemaObject;

  constructor(schemaPath: string | string[]) {
    this.schema =
      typeof schemaPath === 'string'
        ? new ConvertSchemaToObject(schemaPath).run()
        : this.mergeSchemas(schemaPath.map((path) => new ConvertSchemaToObject(path).run()));
  }

  mergeSchemas(schemas: SchemaObject[]) {
    const schema: SchemaObject = { models: [], enums: [] };
    schemas.forEach((s) => {
      schema.models.push(...s.models);
      schema.enums.push(...s.enums);
    });
    return schema;
  }

  async buildSettingsSchema(path = 'adminSettings.json', backAsText?: boolean) {
    const newSchema = mergeSchema(this.schema, path);
    const fileContent = await format(`${JSON.stringify(newSchema)}`, {
      singleQuote: true,
      semi: false,
      trailingComma: 'all',
      parser: 'json',
    });

    !backAsText && writeFileSync(path, fileContent);

    return newSchema;
  }

  async generateGraphql(customOptions?: Omit<Partial<GeneratorOptions>, 'nexusSchema'>) {
    const options: GeneratorOptions = {
      ...defaultOptions,
      ...customOptions,
    };
    return await createGraphql(this.schema, options);
  }

  async generateAdminPages(options?: AdminPagesOptions) {
    const content = options?.pageContent ?? page;
    const generatedText: Record<string, string> = {};
    const filterModels = this.schema.models.filter((model) => !options?.models || options?.models.includes(model.name));

    for (const model of filterModels) {
      const fileContent = await format(content.replace(/#{id}/g, model.name), {
        semi: true,
        trailingComma: 'all',
        singleQuote: true,
        printWidth: 120,
        tabWidth: 2,
        parser: 'babel-ts',
      });
      if (options?.backAsText) {
        generatedText[model.name] = fileContent;
      } else {
        createFile(options?.outPut ?? 'src/pages/admin/models/', `${model.name}.tsx`, fileContent);
      }
    }
    return generatedText;
  }
}

const page = `
import React from 'react';
import PrismaTable from 'components/PrismaTable';

const #{id}: React.FC = () => {
  return <PrismaTable model="#{id}" />;
};

export default #{id};
`;
