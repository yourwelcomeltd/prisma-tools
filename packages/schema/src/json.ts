import { SchemaObject, Model, Enums, Field } from '@paljs/types';
import { PrismaReader } from './PrismaReader';

export class ConvertSchemaToObject extends PrismaReader {
  schemaObject: SchemaObject = {
    models: [],
    enums: [],
    dataSource: undefined,
    generators: [],
  };

  constructor(protected path: string) {
    super(path);
  }

  run() {
    this.getModels();
    this.getEnums();
    this.getDataSource();
    this.getGenerators();

    return this.schemaObject;
  }

  private getModels() {
    if (this.models) {
      for (const model of this.models) {
        const lines = this.blockLines(model);
        const modelObject: Model = {
          name: this.getClassName(lines),
          fields: [],
        };
        let documentation = '';
        for (let i = 1; i + 1 < lines.length; i++) {
          const line = this.lineArray(lines[i]);
          if (line[0].includes('//')) {
            documentation = documentation ? documentation + '\n' + line.join(' ') : line.join(' ');
          } else if (line[0].includes('@@')) {
            modelObject.map = this.getMap(lines[i]);
          } else {
            const type = this.getType(line);
            const field: Field = {
              name: line[0],
              type,
              isId: !!line.find((part) => part.startsWith('@id')),
              unique: !!line.find((part) => part.startsWith('@unique')),
              defaultValue: this.getDefaultValue(lines[i]),
              list: line[1].includes('[]'),
              required: !line[1].includes('[]') && !line[1].includes('?'),
              kind: this.getKind(type),
              documentation,
              map: this.getMap(lines[i]),
            };

            if (field.kind === 'object') {
              field.relation = this.getRelation(lines[i]);
            }
            modelObject.fields.push(field);
            documentation = '';
          }
        }
        modelObject.documentation = this.getModelDocumentation(modelObject.name);
        modelObject.fields
          .filter((item) => item.kind !== 'object')
          .forEach((item) => {
            let relationField = false;
            modelObject.fields
              .filter((field) => field.kind === 'object')
              .forEach((field) => {
                if (!relationField) {
                  relationField = !!field.relation?.fields?.includes(item.name);
                }
              });
            item.relationField = relationField;
          });
        this.schemaObject.models.push({ ...modelObject });
      }
    }
  }

  private getEnums() {
    if (this.enums) {
      for (const item of this.enums) {
        const lines = this.blockLines(item);
        const itemObject: Enums = {
          name: this.getClassName(lines),
          fields: [],
        };
        for (let i = 1; i + 1 < lines.length; i++) {
          const line = this.lineArray(lines[i]);
          !line[0].includes('//') && itemObject.fields.push(line[0]);
        }
        this.schemaObject.enums.push({ ...itemObject });
      }
    }
  }

  private getDataSource() {
    if (this.dataSource) {
      this.schemaObject.dataSource = this.dataSource.map((dataSource) => {
        return {
          provider: dataSource.match(/provider\s+=\s+"(\w+)"/)?.[1] || '',
          url: dataSource.match(/url\s+=\s+(.+)/)?.[1] || '',
        };
      })[0];
    }
  }

  private getGenerators() {
    if (this.generators) {
      this.schemaObject.generators = this.generators.map((generator) => {
        return {
          name: generator.match(/generator\s+(\w+)/)?.[1] || '',
          provider: generator.match(/provider\s+=\s+"(.+)"/)?.[1] || '',
        };
      });
    }
  }
}
