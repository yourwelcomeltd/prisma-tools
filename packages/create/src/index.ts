import spawn from 'cross-spawn';
import chalk from 'chalk';
import { readJSONSync, writeJsonSync, copySync } from 'fs-extra';
import { resolve, join } from 'path';
import { fetchLatestVersionsFor } from './utils/fetch-latest-version-for';
import { log } from '@paljs/display';
import { CliGeneratedExamples } from '@paljs/types';
import * as path from 'path';
import { readdirSync, lstatSync, renameSync, existsSync, mkdirSync, readFileSync, writeFileSync } from 'fs';
import { framework, getPath } from './utils/excludeSettings';

// update to add new FrameWork
export type Frameworks =
  | 'Material UI'
  | 'Material UI + PrismaAdmin UI'
  | 'Tailwind CSS'
  | 'Tailwind CSS + PrismaAdmin UI'
  | 'Chakra UI'
  | 'Chakra UI + PrismaAdmin UI';

export interface AppGeneratorOptions {
  example: CliGeneratedExamples;
  framework?: Frameworks;
  multi?: string | boolean;
  useGit?: string | boolean;
  destinationRoot: string;
  name: string;
  description: string;
  author: string;
  yarn: boolean;
  skipInstall: string | boolean;
  repository: string;
  manager: string;
}

export class AppGenerator {
  constructor(private options: AppGeneratorOptions) {
    this.options.destinationRoot = this.options.name;
    this.options.yarn = this.options.manager === 'yarn';
    this.options.skipInstall = this.options.skipInstall === 'yes';
    this.options.multi = this.options.multi === 'yes';
    this.options.useGit = this.options.useGit === 'yes';
  }

  sourceRoot: string = resolve(__dirname, './examples/', this.options.example);
  packageJson: string[] = [];

  destinationPath(...paths: string[]): string {
    return path.join(this.options.destinationRoot, ...paths);
  }

  loadFiles(path: string) {
    const files = readdirSync(path);
    for (const file of files) {
      if (lstatSync(join(path, file)).isDirectory()) {
        this.loadFiles(join(path, file));
      } else {
        if (file === 'package.json') this.packageJson.push(join(path, file));
        log.info(chalk.green('create ') + file);
      }
    }
  }

  async updatePackages() {
    let fallbackUsed = false;
    for (let i = 0; i < this.packageJson.length; i++) {
      const pkg = readJSONSync(this.packageJson[i]);
      if (i === 0) {
        pkg.name = this.options.name;
        pkg.repository = this.options.repository;
        pkg.author = this.options.author;
        pkg.description = this.options.description;
      }
      for (const dep of ['dependencies', 'devDependencies']) {
        if (pkg[dep]) {
          const { value, isFallback } = await fetchLatestVersionsFor(pkg[dep]);
          pkg[dep] = value;
          if (!fallbackUsed) fallbackUsed = isFallback;
        }
      }

      writeJsonSync(this.packageJson[i], pkg, { spaces: 2 });
    }
    return fallbackUsed;
  }

  async postWrite() {
    const gitInitResult = this.options.useGit
      ? spawn.sync('git', ['init'], {
          stdio: 'ignore',
        })
      : { status: 1 };
    console.log(''); // New line needed
    const spinner = log.spinner(log.withBrand('Retrieving the freshest of dependencies')).start();
    const fallbackUsed = await this.updatePackages();
    if (!fallbackUsed && !this.options.skipInstall) {
      spinner.succeed();

      await new Promise<void>((resolve) => {
        const logFlag = this.options.yarn ? '--json' : '--loglevel=error';
        const cp = spawn(this.options.yarn ? 'yarn' : 'npm', ['install', logFlag], {
          stdio: ['inherit', 'pipe', 'pipe'],
        });

        const getJSON = (data: string) => {
          try {
            return JSON.parse(data);
          } catch {
            return null;
          }
        };

        const spinners: any[] = [];

        if (!this.options.yarn) {
          const spinner = log
            .spinner(log.withBrand('Installing those dependencies (this will take a few minutes)'))
            .start();
          spinners.push(spinner);
        }

        cp.stdout?.setEncoding('utf8');
        cp.stderr?.setEncoding('utf8');
        cp.stdout?.on('data', (data) => {
          if (this.options.yarn) {
            const json = getJSON(data);
            if (json && json.type === 'step') {
              spinners[spinners.length - 1]?.succeed();
              const spinner = log.spinner(log.withBrand(json.data.message)).start();
              spinners.push(spinner);
            }
            if (json && json.type === 'success') {
              spinners[spinners.length - 1]?.succeed();
            }
          }
        });
        cp.stderr?.on('data', (data) => {
          if (this.options.yarn) {
            const json = getJSON(data);
            if (json && json.type === 'error') {
              spinners[spinners.length - 1]?.fail();
              console.error(json.data);
            }
          } else {
            // Hide the annoying Prisma warning about not finding the schema file
            // because we generate the client ourselves
            if (!data.includes('schema.prisma')) {
              console.error(`\n${data}`);
            }
          }
        });
        cp.on('exit', (code) => {
          if (!this.options.yarn && spinners[spinners.length - 1].isSpinning) {
            if (code !== 0) spinners[spinners.length - 1].fail();
            else {
              spinners[spinners.length - 1].succeed();
            }
          }
          resolve();
        });
      });

      const runLocalNodeCLI = (command: string) => {
        if (this.options.yarn) {
          return spawn.sync('yarn', ['run', ...command.split(' ')]);
        } else {
          return spawn.sync('npx', command.split(' '));
        }
      };

      // Ensure the generated files are formatted with the installed prettier version
      const formattingSpinner = log.spinner(log.withBrand('Formatting your code')).start();
      const prettierResult = runLocalNodeCLI('prettier --loglevel silent --write .');
      if (prettierResult.status !== 0) {
        formattingSpinner.fail(
          chalk.yellow.bold("We had an error running Prettier, but don't worry your app will still run fine :)"),
        );
      } else {
        formattingSpinner.succeed();
      }
    } else if (this.options.skipInstall) {
      spinner.succeed();
    } else {
      console.log(''); // New line needed
      spinner.fail(
        chalk.red.bold(
          `We had some trouble connecting to the network, so we'll skip installing your dependencies right now. Make sure to run ${
            this.options.yarn ? "'yarn'" : "'npm install'"
          } once you're connected again.`,
        ),
      );
    }

    if (gitInitResult.status === 0) {
      this.commitChanges();
    } else if (this.options.useGit) {
      log.warning('Failed to run git init.');
      log.warning('Find out more about how to install git here: https://git-scm.com/downloads.');
    }
  }

  commitChanges() {
    const commands: [string, string[], object][] = [
      ['git', ['add', '.'], { stdio: 'ignore' }],
      ['git', ['commit', '-m', 'New paljs app!'], { stdio: 'ignore' }],
    ];
    for (const command of commands) {
      const result = spawn.sync(...command);
      if (result.status !== 0) {
        log.error(`Failed to run command ${command[0]} with ${command[1].join(' ')} options.`);
        break;
      }
    }
  }

  newDir(path: string) {
    !existsSync(path) && mkdirSync(path, { recursive: true });
  }

  excludeMulti(file: string) {
    return (
      (this.options.multi &&
        !['pal.js', 'next.config.js', 'package.json', 'nexusSchema.ts', 'context', 'prisma'].includes(file)) ||
      (!this.options.multi && !file.startsWith('multi_'))
    );
  }

  readDir(path: string) {
    const files = readdirSync(path);
    const frameworkExclude = framework[this.options.framework as Frameworks];
    // update to add new FrameWork
    const withAdmin = !['Tailwind CSS', 'Material UI', 'Chakra UI'].includes(this.options.framework as Frameworks);
    const frameworksFolders = ['material', 'tailwind', 'chakra'].filter((item) => item !== frameworkExclude.folder);
    for (const file of files) {
      if (
        this.excludeMulti(file) &&
        !frameworkExclude.files.includes(file) &&
        !frameworksFolders.includes(file) &&
        file !== '_app'
      ) {
        const newName = file.replace('multi_', '');
        if (lstatSync(join(path, file)).isDirectory()) {
          if (
            !['multi_prisma', frameworkExclude.folder].includes(file) &&
            !path.endsWith('components') &&
            !path.includes(frameworkExclude.folder)
          ) {
            this.newDir(join(this.destinationPath(), getPath(path, this.sourceRoot, frameworkExclude.folder), newName));
          }
          this.readDir(join(path, file));
        } else {
          if (file === '_app.tsx') {
            copySync(
              join(path.replace('pages', '_app'), frameworkExclude.app),
              join(this.destinationPath(), getPath(path, this.sourceRoot, frameworkExclude.folder), newName),
            );
          } else if ((newName === 'pal.config.js' || newName === 'nexusSchema.ts') && !withAdmin) {
            const data = readFileSync(join(path, file), 'utf-8');
            writeFileSync(
              join(this.destinationPath(), getPath(path, this.sourceRoot, frameworkExclude.folder), newName),
              newName === 'nexusSchema.ts'
                ? data.replace('includeAdmin: true,', 'includeAdmin: false,')
                : data.replace(/admin: true,/g, 'admin: false,'),
            );
          } else if (newName === 'package.json') {
            const data = readJSONSync(join(path, file));
            for (const dep of ['dependencies', 'devDependencies']) {
              if (data[dep]) {
                for (const pkgKey in data[dep]) {
                  if (Object.prototype.hasOwnProperty.call(data[dep], pkgKey)) {
                    if (frameworkExclude.packages.includes(pkgKey)) {
                      // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
                      delete data[dep][pkgKey];
                    }
                  }
                }
              }
            }
            writeJsonSync(join(this.destinationPath(), newName), data, {
              spaces: 2,
            });
          } else {
            copySync(
              join(path, file),
              join(this.destinationPath(), getPath(path, this.sourceRoot, frameworkExclude.folder), newName),
            );
          }
        }
      }
    }
  }

  async run() {
    if (this.options.example === 'full-stack-nextjs') {
      this.readDir(this.sourceRoot);
    } else {
      copySync(this.sourceRoot, this.destinationPath());
    }
    process.chdir(this.options.destinationRoot);
    renameSync('gitignore', '.gitignore');
    this.loadFiles('.');
    await this.postWrite();
  }
}
