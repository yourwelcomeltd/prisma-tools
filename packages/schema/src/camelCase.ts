import { writeFileSync } from 'fs';
import { PrismaReader } from './PrismaReader';
import { formatSchema } from '@paljs/utils';
import os from 'os';

export class CamelCase extends PrismaReader {
  async convert(returnString = false): Promise<string | void> {
    let newData = this.data;
    if (this.models) {
      for (const model of this.models) {
        const lines = this.blockLines(model);
        let modelName = '';
        const fieldsList: {
          old: string;
          new: string;
        }[] = [];
        for (let i = 0; i + 1 < lines.length; i++) {
          const line = this.lineArray(lines[i]);
          if (i === 0) {
            modelName = line[1];
          }
          // check if we are in the first line that has the model name
          // check if we need to change this model name
          if (i === 0 && CamelCase.checkIfModelNameNeedToConvert(line[1])) {
            // add a map with the old model name
            lines.splice(1, 0, `@@map("${modelName}")`);
            // get the new model name converted to the camel case
            line[1] = this.convertWord(line[1], true);
            // update the model name in the original string
            lines[0] = line.join(' ');
          }
          // check if this line is a field not an attributes or a comment
          else if (!lines[i].includes('@@') && !lines[i].includes('//')) {
            const kind = this.getKind(this.getType(line));
            const fieldName = line[0];
            const type = line[1];
            if (fieldName.includes('_')) {
              if (kind !== 'object') line.push(`@map("${fieldName}")`); // keep the original field name for @map
              line[0] = this.convertWord(fieldName);
              fieldsList.push({
                old: fieldName,
                new: line[0],
              });
            }
            if (CamelCase.checkIfModelNameNeedToConvert(type) && kind === 'object') {
              line[1] = this.convertWord(type, true);
            }
            lines[i] = line.join(' ');
          }
        }
        for (const field of fieldsList) {
          const fieldPattern = new RegExp(`\\b${field.old}\\b`, 'g'); // use word boundaries to match whole words only
          lines.forEach((line, index) => {
            if (!line.endsWith(`@map("${field.old}")`)) {
              lines[index] = line.replace(fieldPattern, field.new);
            }
          });
        }
        const newModel = lines.join(`
`);

        const pattern = new RegExp(`model ${modelName}[\\s\\S]*?}\n`);
        newData = newData.replace(pattern, newModel);
      }

      let [[, output]] = await formatSchema({
        schemas: [['schema.prisma', newData]],
      });

      output = output.trimEnd() + os.EOL;

      if (returnString) {
        return output;
      }

      writeFileSync(this.path, output);
    }
  }

  // convert word to camel case
  private convertWord(word: string, model = false) {
    return word
      .split('_')
      .map((item, index) =>
        !model && index === 0 ? item.toLowerCase() : item.charAt(0).toUpperCase() + item.slice(1).toLowerCase(),
      )
      .join('');
  }

  // check is the word start with a lower case letter
  private static checkFirstLetterLowerCase(word: string) {
    return word[0] === word[0].toLowerCase();
  }

  /*
   * check if the model name starts with a lower case or has an underscore
   */
  private static checkIfModelNameNeedToConvert(modelName: string) {
    return modelName.includes('_') || CamelCase.checkFirstLetterLowerCase(modelName);
  }
}
