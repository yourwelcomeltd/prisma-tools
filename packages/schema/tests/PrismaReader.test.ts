import { PrismaReader } from '../src';
import { join } from 'path';

describe('test the Prisma Reader class', () => {
  const prismaReader = new PrismaReader(join(__dirname, './schemas/schema.prisma'));

  test('read prisma models', () => {
    expect(prismaReader.models).toMatchSnapshot();
  });

  test('read prisma enums', () => {
    expect(prismaReader.enums).toMatchSnapshot();
  });

  test('read prisma model documentation', () => {
    expect(prismaReader.getModelDocumentation('User')).toMatchSnapshot();
  });

  test('read prisma datasource', () => {
    expect(prismaReader.dataSource).toMatchSnapshot();
  });

  test('read prisma generators', () => {
    expect(prismaReader.generators).toMatchSnapshot();
  });

  test('read prisma datasource with url string', () => {
    const prismaReader = new PrismaReader(join(__dirname, './schemas/schema-datasource-with-url-string.prisma'));

    expect(prismaReader.dataSource).toMatchInlineSnapshot(`
      [
        "datasource db {
        provider = "postgresql"
        url      = "file:./dev.db"
      }
      ",
      ]
    `);
  });
});
