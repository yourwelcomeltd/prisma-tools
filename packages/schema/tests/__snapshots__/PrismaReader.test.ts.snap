// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test the Prisma Reader class read prisma datasource 1`] = `
[
  "datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}
",
]
`;

exports[`test the Prisma Reader class read prisma enums 1`] = `
[
  "
enum Role {
  ADMIN
  USER
}
",
]
`;

exports[`test the Prisma Reader class read prisma generators 1`] = `
[
  "
generator client {
  provider = "prisma-client-js"
}
",
  "
generator pothos {
  provider = "prisma-pothos-types"
}
",
]
`;

exports[`test the Prisma Reader class read prisma model documentation 1`] = `"// User modal documentation"`;

exports[`test the Prisma Reader class read prisma models 1`] = `
[
  "
model User {
  id          Int      @id @default(autoincrement())
  createdAt   DateTime @default(now())
  email       String   @unique
  name        String?
  // password documentation
  // password second line documentation
  password    String
  permissions Json     @default("{}")
  posts       Post[]
}
",
  "
model Post {
  id        Int      @id @default(autoincrement())
  published Boolean  @default(false)
  title     String
  author    User?    @relation(fields: [authorId], references: [id])
  authorId  Int?
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
}
",
]
`;
